In [170]:
import pandas as pd
df = pd.read_csv(r'C:\Users\91987\Downloads\python\bipp-datasets\projects\fertilizer-mis\data\interim\dealer_data.csv',low_memory=False)

In [ ]:
df.head()

In [171]:
index_list=[]
for i in range(1,360335):
    index_list.append(i)
df.insert(0,"Index",index_list,True)
#df.set_index("Index")   
df.head()

Index                   State Name             District Name Dealer Id  \
0      1  Andaman and Nicobar Islands                  Nicobars    658932   
1      2  Andaman and Nicobar Islands                  Nicobars    937065   
2      3  Andaman and Nicobar Islands                  Nicobars    937067   
3      4  Andaman and Nicobar Islands                  Nicobars   1206716   
4      5  Andaman and Nicobar Islands  North And Middle Andaman    936262   

  Dealer Type              Agency Name Correct Agency Name  Owner Name  \
0    RETAILER                    trnAN                 NaN         NaN   
1    RETAILER   SUB DEPOT GOVIND NAGAR                 NaN         NaN   
2    RETAILER   SUB DEPOT GANDHI NAGAR                 NaN         NaN   
3  WHOLESALER  Main Store Campbell Bay                 NaN         NaN   
4    RETAILER         SUB DEPOT RANGAT                 NaN         NaN   

  Existing Mobile  New Mobile if any  ... Original Dealer Id  \
0      9434262481                NaN  ...                NaN   
1      8900905069                NaN  ...                NaN   
2      9476023727                NaN  ...                NaN   
3      9474234749                NaN  ...                NaN   
4      8900908157                NaN  ...                NaN   

  Sub District Name Village Name           Dealer Address Line 1  \
0    Baganpara (Pt)     Amingaon                             NaN   
1     Great Nicobar          NaN                             NaN   
2     Great Nicobar          NaN                             NaN   
3     Great Nicobar          NaN  Near Zonal Agricultural Office   
4            Rangat          NaN                             NaN   

   Dealer Address Line 2 Dealership Nature Name        Dealer License  \
0                    NaN       Private Retailer           nictest2017   
1                    NaN                  Other  DOA GovindNagar Depo   
2                    NaN                  Other  DOA Gandhinagar Depo   
3                    NaN                  Other            ANI/DOA/56   
4                    NaN                  Other      DOA Rangat Depot   

    License Issue Date  License Expiry Date Unnamed: 22  
0  2017-03-10 00:00:00  2020-03-10 00:00:00         NaN  
1                  NaN                  NaN         NaN  
2                  NaN                  NaN         NaN  
3  2020-01-17 00:00:00  2025-01-17 00:00:00         NaN  
4                  NaN                  NaN         NaN  

[5 rows x 24 columns]

In [ ]:
df.isnull().sum()

In [ ]:
df['Dealer Id'].value_counts()

In [ ]:
df['State Name'].value_counts()

In [172]:
import os
import csv
def read_folder(csv_folder):
    files = os.listdir(csv_folder)
    state_names = []
    for f in files:
      state_names.append(f)
    return(state_names)
states=read_folder(r"C:\Users\91987\Downloads\python\bipp-datasets\projects\fertilizer-mis\data\raw\2_dealer_data")
statenames=[]
for s in states:
    s=s.replace('.csv',"")
    #print(s)
    statenames.append(s)
print(statenames)

['Andaman and Nicobar Islands', 'Andhra Pradesh', 'Arunachal Pradesh', 'Assam', 'Bihar', 'Chandigarh', 'Chhattisgarh', 'Dadra and Nagar Haveli', 'Daman and Diu', 'Delhi', 'Goa', 'Gujarat', 'Haryana', 'Himachal Pradesh', 'Jammu and kashmir', 'Jharkhand', 'Karnataka', 'Kerala', 'Lakshadweep', 'Madhya Pradesh', 'Maharashtra', 'Manipur', 'Meghalaya', 'Mizoram', 'Nagaland', 'Odisha', 'Puducherry', 'Punjab', 'Rajasthan', 'Sikkim', 'Tamil Nadu', 'Telangana', 'Tripura', 'Uttar Pradesh', 'Uttarakhand', 'West Bengal']


In [173]:
unclean_data=[]
for ind,row in df.iterrows():
    found=0
    for s in statenames:
        if row['State Name']==s:
            found=1
    
    if found!=1:
        wrong_value = row['State Name']
        unclean_data.append(row)
        df.drop(index=ind,inplace=True)

In [174]:
df.tail(10)

Index   State Name District Name Dealer Id Dealer Type  \
360324  360325  West Bengal       Purulia   1210842    RETAILER   
360325  360326  West Bengal       Purulia   1217594    RETAILER   
360326  360327  West Bengal       Purulia   1217600    RETAILER   
360327  360328  West Bengal       Purulia   1217614    RETAILER   
360328  360329  West Bengal       Purulia   1217618    RETAILER   
360329  360330  West Bengal       Purulia   1219893    RETAILER   
360330  360331  West Bengal       Purulia   1219895    RETAILER   
360331  360332  West Bengal       Purulia   1219901    RETAILER   
360332  360333  West Bengal       Purulia   1219903    RETAILER   
360333  360334  West Bengal       Purulia   1226615    RETAILER   

                 Agency Name Correct Agency Name  Owner Name Existing Mobile  \
360324    JHALDA LSAPMCS LTD                 NaN         NaN    9932186264.0   
360325    MAHESWAR PARAMANIK                 NaN         NaN    9635679113.0   
360326          GOPAL MAHATO                 NaN         NaN    7908628007.0   
360327          KAJAL MANDAL                 NaN         NaN    9635389163.0   
360328         AVIJIT MANDAL                 NaN         NaN    7602363799.0   
360329  SWADHIN KUMAR MAHATO                 NaN         NaN    9064449326.0   
360330          NANDALAL DAS                 NaN         NaN    9531568844.0   
360331     NABIN KUMAR MAJHI                 NaN         NaN    6295471723.0   
360332         HARIPADA PAUL                 NaN         NaN    7797506237.0   
360333        SULPANI MAHATO                 NaN         NaN    9932000705.0   

        New Mobile if any  ... Original Dealer Id Sub District Name  \
360324                NaN  ...                NaN       Purulia - I   
360325                NaN  ...                NaN       Purulia - I   
360326                NaN  ...                NaN       Jhalda - II   
360327                NaN  ...                NaN              Hura   
360328                NaN  ...                NaN          Kashipur   
360329                NaN  ...                NaN     Manbazar - II   
360330                NaN  ...                NaN           Bundwan   
360331                NaN  ...                NaN             Arsha   
360332                NaN  ...                NaN     Manbazar - II   
360333                NaN  ...                NaN          Kashipur   

       Village Name Dealer Address Line 1  Dealer Address Line 2  \
360324          NaN          PURULIA TOWN                    NaN   
360325          NaN                BHADSA                    NaN   
360326          NaN            BEGUNKODAR                    NaN   
360327          NaN              KALABANI                    NaN   
360328          NaN                ISANDA                    NaN   
360329          NaN               CHOUKAN                   BARI   
360330          NaN                DHADKA                 DHADKA   
360331          NaN            UPARJAMBAD                AHARRAH   
360332          NaN                  BARI                   BARI   
360333          NaN           NARAYANGARH                    NaN   

       Dealership Nature Name        Dealer License License Issue Date  \
360324                   PACS  627/ADA(Admn.)/RNPUR   06-02-2019 00:00   
360325       Private Retailer                F33387   04-12-2020 00:00   
360326       Private Retailer           PRLS/R-1018   01-08-2019 00:00   
360327       Private Retailer                F22456   05-02-2020 00:00   
360328       Private Retailer                F24812   10-06-2020 00:00   
360329       Private Retailer            PRLS/R-989   01-11-2018 00:00   
360330       Private Retailer            PRLS/R-959   01-09-2018 00:00   
360331       Private Retailer                F19504   01-02-2019 00:00   
360332       Private Retailer            PRLS/R-968   01-09-2018 00:00   
360333       Private Retailer                F30372   18-09-2020 00:00   

       License Expiry Date Unnamed: 22  
360

In [ ]:
df['State Name'].value_counts()

original df has 0 indexing, 0 to 360333
i created index column with 1 indexing, 1 to 360334
i dropped some rows from df and put them in unclean data
so now shape should show 360203,24
there should be 131 rows in unclean data

In [175]:
df.shape

(360203, 24)

In [ ]:
df.isnull().sum()

In [ ]:
df.head()

In [ ]:
df.tail()

In [ ]:
#print(unclean_data)

In [176]:
unclean_df=pd.DataFrame(unclean_data)
print(unclean_df)

         Index                      State Name District Name  \
356132  356133                  Dist.   Nadia.           NaN   
356152  356153                         Arbandi           NaN   
356167  356168                P.S.   Dhantala.           NaN   
356197  356198  P.S.   Chakdaha Dist.   Nadia.           NaN   
356199  356200  P.S.   Chakdaha Dist.   Nadia.           NaN   
...        ...                             ...           ...   
356732  356733  P.S.   Chakdaha Dist.   Nadia.           NaN   
356733  356734                         CK/2007           NaN   
356735  356736               P.O.   Mollabelia           NaN   
356736  356737               P.S.   Haringhata           NaN   
356737  356738                            HG/0           NaN   

               Dealer Id                     Dealer Type Agency Name  \
356132  Private Retailer                RN005/NAD/200912         NaN   
356152  Private Retailer      Rana/347/88/Retail/SNT/ 10         NaN   
356167  Private

unclean df- shape should be 131,24

In [177]:
unclean_df.shape

(131, 24)

In [ ]:
unclean_df.isnull().sum()

output of shape of unclean df should now be 131,8

In [178]:
for c in unclean_df.columns:
    if unclean_df[c].count()==0:
        unclean_df=unclean_df.drop(columns=c)
print(unclean_df.shape)

(131, 8)


In [180]:
unclean_df.to_excel(r'C:\Users\91987\Downloads\python\unclean_data.xlsx')

In [ ]:
#for index,row in unclean_df.iterrows():
#    if row["Dealer Id"]=="Private Wholesaler" or row["Dealer Id"]=="Private Retailer":
#        df.set_value(,'Dealership Nature Name',row['Dealer Id'],)

In [181]:
df["Dealership Nature Name"].count()

360148

In [182]:
indexofdflist=[]
indexofdflist=pd.DataFrame(df.index)

In [183]:
indexofdflist

0
0            0
1            1
2            2
3            3
4            4
...        ...
360198  360329
360199  360330
360200  360331
360201  360332
360202  360333

[360203 rows x 1 columns]

In [ ]:
df.shape

In [ ]:
df["Dealership Nature Name"].value_counts()

In [184]:
found =0
dealer_id_list=[]
index_list_df=[]
for index1,row in df.iterrows():
    if row["Dealership Nature Name"]=="Private Retailer" or row["Dealership Nature Name"]=="PACS" or row["Dealership Nature Name"]=="Private Wholsaler" or row["Dealership Nature Name"]=="District Cooperative Marketing Society" or row["Dealership Nature Name"]=="State Marketing Federation" or row["Dealership Nature Name"]=="Other" or row["Dealership Nature Name"]=="Manufacturer Outlet" or row["Dealership Nature Name"]=="Marketer" or row["Dealership Nature Name"]=="Mixture Manufacturer" or row["Dealership Nature Name"]=="institutional_buyer" or row["Dealership Nature Name"]=="HUB Retailer":
        found+=1
    else:
        dealer_id_list.append(row["Dealer Id"])
        index_list_df.append(row["Index"])
        #for index2,rows in unclean_df.iterrows():
        #if rows["Index"]==(row["Index"]+1) and (rows["Dealer Id"]=="Private Wholesaler" or rows["Dealer Id"]=="Private Retailer")
        #row["Dealership Nature Name"]=rows["Dealer Id"]
        #df["Dealership Nature Name"][index1]=row["Dealer Id"]
        #df.at[index1,"Dealership Nature Name"]=row["Dealer Id"]
        #print(df["Dealership Nature Name"][index1])
#print(found)
print(dealer_id_list)
print(index_list_df)
#unclean_df

['275228', '340740', '340805', '340848', '340967', '340968', '340978', '341053', '341060', '341065', '341284', '341324', '341334', '341537', '341554', '341555', '341580', '341600', '341615', '341781', '341838', '341869', '341947', '341950', '341993', '341998', '342027', '342042', '342069', '342080', '342109', '342112', '342116', '342120', '342153', '342200', '342587', '342592', '342595', '342596', '342600', '342619', '342621', '342626', '342627', '342640', '342643', '342646', '342648', '342685', '356113', '356134', '356241', '356284', '356285']
[78137, 356132, 356152, 356167, 356197, 356199, 356203, 356217, 356220, 356224, 356261, 356270, 356277, 356321, 356332, 356334, 356342, 356349, 356359, 356393, 356407, 356414, 356429, 356433, 356448, 356453, 356462, 356469, 356480, 356488, 356500, 356503, 356509, 356514, 356525, 356532, 356556, 356560, 356563, 356567, 356571, 356578, 356581, 356586, 356590, 356598, 356603, 356607, 356611, 356631, 356695, 356701, 356716, 356732, 356735]


In [ ]:
df.tail(10)

In [ ]:
df.shape

In [ ]:
df["Dealership Nature Name"].count()

In [ ]:
unclean_df

In [ ]:
#print(df.at[356199,"Dealership Nature Name"])
#df.at[356199,"Dealer License"]

In [ ]:
df["Dealership Nature Name"].isnull().sum()

In [ ]:
df.shape

In [ ]:
unclean_df["Dealer Id"].value_counts()

In [ ]:
unclean_df["Dealer Type"].value_counts()

In [ ]:
df["Dealer License"].count()

In [ ]:
df["Dealer License"].isnull().sum()

In [ ]:
#somehow not working.. i wanted to find indices where dealer license is null
# check_nan=pd.DataFrame(df['Dealer License'].isnull())
# #check_nan = df['Dealer License'].isnull()
# print(check_nan)
# check_nan_list=[]
# for index,rows in check_nan.iterrows():
#     if rows["Dealer License"]=="True":
#         check_nan_list.append(index)
#         print(index)
# print(check_nan_list)

In [ ]:
#nested loop takes time
# for index1,r in df.iterrows():
#     #check_nan = df['Dealer License'].isnull().values.any()
#     #if r["Dealer License"].isnan():
#     for index,row in unclean_df.iterrows():
#         if row["Dealer Type"]!="9434553430" and row["Dealer Type"]!="7478926270":
#             if r["Index"]==row["Index"]-1:
#                 df["Dealer License"][index1]=row["Dealer Type"]
# df["Dealer License"].count()

In [ ]:
df.index

In [ ]:
df.set_index("Index")

In [ ]:
unclean_df

In [ ]:
df["Dealer License"].count()

In [ ]:
for index,row in unclean_df.iterrows():
    if row["Dealer Type"]!="9434553430" and row["Dealer Type"]!="7478926270":
        #print(row["Dealer Type"])
        #df["Dealer License"][row["Index"]-1]=row["Dealer Type"]
        ind_df=row["Index"]-1
        df.at[ind_df,"Dealer License"]=row["Dealer Type"]
df["Dealer License"].count()

In [ ]:
unclean_df
df.at[356199,"Dealer License"]

In [ ]:
"""
8 columns in unclean data
7 columns of concern
c
d
e-error
f-some error
g
h
i-some error
"""

In [ ]:
unclean_df["Duplicate Dealer(Yes/No) if Yes Then enter Original Id"].value_counts()

In [ ]:
df.shape

In [ ]:
df=df.drop(columns="Address")

In [ ]:
df.head()

In [ ]:
index_list=[]
for i in range(1,360333):
    index_list.append("temp")
df.insert(0,"Address",index_list,True)
df.head()

In [ ]:
df.index

In [ ]:
df.set_index("Index")

In [ ]:
for index,row in unclean_df.iterrows():
    ind_df=row["Index"]-1
    if row["Duplicate Dealer(Yes/No) if Yes Then enter Original Id"]=="P.O. + Vill.   Gangnapur" or row["Duplicate Dealer(Yes/No) if Yes Then enter Original Id"]=="Vill.   Mithapur":
        #df["Address"][ind_df]=row["Duplicate Dealer(Yes/No) if Yes Then enter Original Id"]
        #print(df["Address"][ind_df])
        print(ind_df)